Bottom: d05239b3803616427bde5025c139757d9e5d34b5
Top:    05bd0567ea81da317bc4316ff2fdc38d17d70bbf
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-04-17 07:44:14 +0800

tuntap: enable premmption early

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index a1ba262..16de8ab 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1692,6 +1692,8 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 			goto err_xdp;
 		}
 	}
+	rcu_read_unlock();
+	preempt_enable();
 
 	skb = build_skb(buf, buflen);
 	if (!skb) {
@@ -1705,9 +1707,6 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	get_page(alloc_frag->page);
 	alloc_frag->offset += buflen;
 
-	rcu_read_unlock();
-	preempt_enable();
-
 	return skb;
 
 err_redirect:
